kimchi/src/tests/foreign_field_mul.rs

```diff
@@ -0,0 +1,1409 @@
+            TestFramework::<G>::default()
+                .gates(gates.clone())
+                .witness(witness.clone())
+                .lookup_tables(vec![foreign_field_mul::gadget::lookup_table()])
+                .foreign_modulus(Some(foreign_field_modulus.clone()))
+                .setup(),
+        )
+    } else {
+        None
+    };
+
+    let cs = if let Some(runner) = runner.as_ref() {
+        runner.prover_index().cs.clone()
+    } else {
+        // If not full mode, just create constraint system (this is much faster)
+        ConstraintSystem::create(gates.clone())
+            .foreign_field_modulus(&Some(foreign_field_modulus.clone()))
+            .build()
+            .unwrap()
+    };
+
+    // Perform witness verification that everything is ok before invalidation (quick checks)
+    for (row, gate) in gates.iter().enumerate().take(witness[0].len()) {
+        let result = gate.verify_witness::<G>(row, &witness, &cs, &witness[0][0..cs.public]);
+        if result.is_err() {
+            return (result, witness);
+        }
+    }
+
+    if let Some(runner) = runner {
+        // Perform full test that everything is ok before invalidation
+        runner.prove_and_verify::<EFqSponge, EFrSponge>();
+    }
+
+    if !invalidations.is_empty() {
+        for ((row, col), mut value) in invalidations {
+            // Invalidate witness
+            if witness[col][row] == value {
+                // If the invalidation would be the same value choose a different one
+                // Don't let it wrap around
+                assert_ne!(value, G::ScalarField::zero());
+                value -= G::ScalarField::one();
+            }
+            witness[col][row] = value;
+        }
+
+        if !target_plookups {
+            // Check witness verification fails
+            // When targeting the plookup constraints the invalidated values would cause custom constraint
+            // failures, so we want to suppress these witness verification checks when doing plookup tests.
+            for (row, gate) in gates.iter().enumerate().take(witness[0].len()) {
+                let result =
+                    gate.verify_witness::<G>(row, &witness, &cs, &witness[0][0..cs.public]);
+                if result.is_err() {
+                    return (result, witness);
+                }
+            }
+        }
+
+        // Catch plookup failures caused by invalidation of witness
+        if full {
+            TestFramework::<G>::default()
+                .gates(gates.clone())
+                .witness(witness.clone())
+                .lookup_tables(vec![foreign_field_mul::gadget::lookup_table()])
+                .foreign_modulus(Some(foreign_field_modulus.clone()))
+                .setup()
+                .prove_and_verify::<EFqSponge, EFrSponge>();
+        }
+    }
+
+    (Ok(()), witness)
+}
+
+/// Generate a random foreign field element x whose addition with the negated foreign field modulus f' = 2^t - f results
```

Comment:
I think the proof shouldn't be placed here but in the RFC

Reply:
Put into RFC in opt2 pr